reactos/drivers/usb/usbport/iface.c

```diff
@@ -0,0 +1,916 @@
+        HcDisable = 1;
+    }
+
+    Status = USBPORT_SetRegistryKeyValue(FdoExtension->CommonExtension.LowerPdoDevice,
+                                         (HANDLE)1,
+                                         REG_DWORD,
+                                         L"HcDisableSelectiveSuspend",
+                                         &HcDisable,
+                                         sizeof(HcDisable));
+
+    if (NT_SUCCESS(Status))
+    {
+        if (Enable) 
+            FdoExtension->Flags |= USBPORT_FLAG_SELECTIVE_SUSPEND;
+        else
+            FdoExtension->Flags &= ~USBPORT_FLAG_SELECTIVE_SUSPEND;
+    }
+
+    return Status;
+}
+
+NTSTATUS
+USB_BUSIFFN
+USBHI_GetExtendedHubInformation(IN PVOID BusContext,
+                                IN PDEVICE_OBJECT HubPhysicalDeviceObject,
+                                IN OUT PVOID HubInformationBuffer,
+                                IN ULONG HubInfoLen,
+                                IN OUT PULONG LenDataReturned)
+{
+    PDEVICE_OBJECT PdoDevice;
+    PUSBPORT_RHDEVICE_EXTENSION PdoExtension;
+    PDEVICE_OBJECT FdoDevice;
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+    PUSBPORT_REGISTRATION_PACKET Packet;
+    ULONG NumPorts;
+    ULONG ix;
+    PUSB_EXTHUB_INFORMATION_0 HubInfoBuffer;
+    ULONG PortStatus = 0;
+    ULONG PortAttrX;
+
+    DPRINT("USBHI_GetExtendedHubInformation: ... \n");
+
+    PdoDevice = (PDEVICE_OBJECT)BusContext;
+    PdoExtension = (PUSBPORT_RHDEVICE_EXTENSION)PdoDevice->DeviceExtension;
+    FdoDevice = PdoExtension->FdoDevice;
+    FdoExtension = (PUSBPORT_DEVICE_EXTENSION)FdoDevice->DeviceExtension;
+    Packet = &FdoExtension->MiniPortInterface->Packet;
+
+    HubInfoBuffer = (PUSB_EXTHUB_INFORMATION_0)HubInformationBuffer;
+
+    if (HubPhysicalDeviceObject != PdoDevice)
+    {
+        *LenDataReturned = 0;
+        return STATUS_NOT_SUPPORTED;
+    }
+
+    if (HubInfoLen < sizeof(USB_EXTHUB_INFORMATION_0))
+    {
+        *LenDataReturned = 0;
+        return STATUS_BUFFER_TOO_SMALL;
+    }
+
+    NumPorts = PdoExtension->RootHubDescriptors->Descriptor.bNumberOfPorts;
+    HubInfoBuffer->NumberOfPorts = NumPorts;
+
+    if (NumPorts == 0)
+    {
+        *LenDataReturned = sizeof(USB_EXTHUB_INFORMATION_0);
+        return STATUS_SUCCESS;
+    }
+
+    for (ix = 1; ix <= HubInfoBuffer->NumberOfPorts; ++ix)
+    {
+        HubInfoBuffer->Port[ix].PhysicalPortNumber = ix;
+        HubInfoBuffer->Port[ix].PortLabelNumber = ix;
+        HubInfoBuffer->Port[ix].VidOverride = 0;
+        HubInfoBuffer->Port[ix].PidOverride = 0;
+        HubInfoBuffer->Port[ix].PortAttributes = 0;
+
+        if (Packet->MiniPortFlags & USB_MINIPORT_FLAGS_USB2)
+        {
+            HubInfoBuffer->Port[ix].PortAttributes = USB_PORTATTR_SHARED_USB2;
+
+            Packet->RH_GetPortStatus(FdoExtension->MiniPortExt,
+                                     ix,
+                                     &PortStatus);
+
+            if (PortStatus & 0x8000)
```

Comment:
It seems this is an undocumented feature of Port Status (USB 2.0 Spec .: 11.24.2.7.1 Port Status Bits), Reserved: Bit 13-15. It corresponds to the structure _USB_PORT_STATUS in usbmport.h, field USHORT Reserved2: 3
The Port Owner status (bit 13) (EHCI Spec .: Table 2-16 .PORTSC Port Status and Control) is transmitted via PortStatus for USB_EXTPORT_INFORMATION_0 :: PortAttributes (USB_PORTATTR_OWNED_BY_CC).
While I'm not sure about this at 100%, because it can only be checked on HW.